### PR TITLE
Implementation - Find and Fill Method for Dropout Layer

### PR DESCRIPTION
**Pull request by MarkFischinger**
Following our discussion in issue #3662, I've implemented the new algorithm in the dropout layer.

### CODE REVIEW

**Review by shrit — APPROVED** [review]

**Review by rcurtin — CHANGES_REQUESTED** [review]
Should we split this one into two implementations, like #3683?  The numbers in #3662 show that this implementation is actually a slowdown, although the implementation here is general enough to be used with Bandicoot too.

(Requesting changes just because I don't want to merge a slowdown without a bit of discussion)

**Comment by shrit** [comment]
it is up to you, usually the dropout is happening once to my understanding per epoch, while there is a slow down, but it is not going to affect the general speed of training.
We can do the split, but I would trade the code maintenance and size compared to the number of nanoseconds we will lose :smile:

**Comment by rcurtin** [comment]
It will happen every step of the forward pass, so the slowdown may be noticeable.  One quick test would just be to run the `mnist_simple` and/or `mnist_cnn` examples with both implementations and see what the overall time per epoch is.

**Comment by MarkFischinger** [comment]
@rcurtin @shrit Absolutely, I'm on the same page. I was actually leaning towards running additional benchmarks for that exact reason before. I've got a few ideas to potentially optimize the algorithm, and I'm eager to test them out in some upcoming benchmarks. Fingers crossed, I should be able to run them later today :)

**Comment by MarkFischinger** [comment]
@rcurtin @shrit The new algorithm appears slightly faster than the old one when processing small matrices but significantly slower with larger matrices (also visible in the graph in the discussion before). If it's not necessary to replace it, I wouldn't. However, if we need to eliminate the transform function, this could be a viable alternative.

**For mnist_simple:**

    New Algorithm:
        Duration: 78910ms
        Loss: 0.0331713
    Old Algorithm:
        Duration: 79709ms
        Loss: 0.0331713

**For mnist_cnn:**

    Old Algorithm:
        Duration: 403866ms
        Loss: 0.919632 (Epoch: 32.9s, Step: 170ms)
        Validation loss: 2894.57
    New Algorithm:
        Duration: 415430ms
        Loss: 0.919632 (Epoch: 31.93s, Step: 174ms)
        Validation loss: 2894.57
        Accuracy: Train = 86.9947%, Valid = 86.5476%

**Comment by shrit** [comment]
@MarkFischinger Okay, I see the benchmark, in this case it would make sense to have both implementation. Please update this PR by following a similar structure that is done in this #3683

**Comment by MarkFischinger** [comment]
@shrit In the latest commit, I've implemented both algorithms using a structure similar to what you did in #3683. I have a question about that: Why do we wrap the entire function within the #ifndef directive? Could it be placed inside the function instead? Is there a specific reason for this that I might not know about yet? I'd appreciate understanding your reasoning behind that :)

**Comment by shrit** [comment]
@MarkFischinger the other implementation that you added is for bandicoot library. It is a linear algebra library that we are developing (it has the same API as armadillo) to provide GPU acceleration for matrices operations.
The idea of putting all of that in a `#ifdef` is to have no dependency on bandicoot, otherwise, you will not be able to run mlpack or the tests without including bandicoot header which is something we are trying to avoid.
Technically, in the future, the user only needs to define `MLPACK_HAS_COOT` and then the makefile will look for bandicoot.
Our objective is too not add dependencies to mlpack, since they add a massive burden regarding maintenance and compilation, and when necessary we are only adding single file header only dependencies.

**Comment by MarkFischinger** [comment]
@shrit Thanks for your response! :) Sorry if my initial question wasn’t as clear. I've understood the general concept, but I'm specifically curious about the decision to place the `ifndef` **outside** the function. Is this meant to make it easier to distinguish between GPU and CPU operations, thereby improving maintainability?

Because my first thought was to implement it directly **inside** like this (for #3683 for example):
```cpp
template<typename MatType>
void LogSoftMaxType<MatType>::ForwardImpl(const MatType& input,
                                          MatType& output)
{
  MatType maxInput = repmat(max(input), input.n_rows, 1);
  output = (maxInput - input);

  ApplyExponentialFunction(output);

  maxInput.each_row() += log(sum(output));
  output = input - maxInput;
}

template<typename MatType>
void LogSoftMaxType<MatType>::ApplyExponentialFunction(MatType& output)
{
#ifdef MLPACK_HAS_COOT // <-- Inside the function like this 
  if constexpr (coot::is_coot_type<MatType>::value)
  {
    output = exp(output * -1);
  }
  else
#endif
  {
    // Approximation of the base-e exponential function. Credits to Leon Bottou.
    output.transform([](double x)
    {
      static constexpr double A0 = 1.0, A1 = 0.125, A2 = 0.0078125, A3 = 0.00032552083, A4 = 1.0172526e-5;
      if (x < 13.0)
      {
        double y = A0 + x * (A1 + x * (A2 + x * (A3 + x * A4)));
        y *= y; y *= y; y *= y; return 1 / y;
      }
      return 0.0;
    });
  }
}
```

I couldn't find any best practices in C++ on this topic on the web. Do you know which is the preferred implementation?

**Comment by shrit** [comment]
@MarkFischinger we can not use `if constexpr` because we do not support yet C++17

**Review by shrit — APPROVED** [review]

**Review by rcurtin — COMMENTED** [review]
Looking good, just some comments on the multiple implementations whenever you have a chance. 👍

File: src/mlpack/methods/ann/layer/dropout_impl.hpp [hunk 1]
```diff
@@ -92,6 +100,28 @@ void DropoutType<MatType>::Forward(const MatType& input, MatType& output)
...
+  if (!this->training)
+  {
+    output = input;
+  }
+  else
+  {
+    mask.randu(input.n_rows, input.n_cols);
+    arma::uvec indices = arma::find(mask > ratio);
+    mask.zeros();
+    mask.elem(indices).fill(1);
```
[the remark this answers is not in the document]
Comment (MarkFischinger) [reply src/mlpack/methods/ann/layer/dropout_impl.hpp]: @rcurtin You can take a look at my implementation in the latest commit. It should be a bit faster and is fully functional. I avoided using arma:: to keep it more general.

File: src/mlpack/methods/ann/layer/dropout_impl.hpp [hunk 2]
```diff
@@ -74,10 +74,18 @@ DropoutType<MatType>::operator=(DropoutType&& other)
   return *this;
 }
 
+
```
[the remark this answers is not in the document]
Comment (shrit) [reply src/mlpack/methods/ann/layer/dropout_impl.hpp:77]: Also, this one is still missing.

File: src/mlpack/methods/ann/layer/dropout.hpp [hunk 3]
```diff
@@ -77,6 +77,24 @@ class DropoutType : public Layer<MatType>
...
+  void ForwardImpl(const T& input, T& output);
+
+  /**
+   * General implementation of the forward pass of the dropout layer.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   */
+  template<typename T = MatType, typename std::enable_if_t<!arma::is_arma_type<T>::value, int> = 0>
+  void ForwardImpl(const T& input, T& output);
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/dropout.hpp:96]: @MarkFischinger
Looks better, just use directly `MatType` instead of `T`.  It is better for readability because `T`  does not mean a lot.
Also, could you break the line? mlpack code base is usually under 80 lines of code.
Also, the syntax is correct, I would remove `int` and instead add a `*` after `>` so the  signature looks as follows:
```c++
typename std::enable_if_t<!arma::is_arma_type<T>::value>* = 0>
```
The reason for this is to have the same signature over all the code base, making it easier for everyone.

File: src/mlpack/methods/ann/layer/dropout_impl.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rcurtin) [reply src/mlpack/methods/ann/layer/dropout_impl.hpp]: Nice, it looks good to me! Thanks for doing that.

File: src/mlpack/methods/ann/layer/dropout_impl.hpp [hunk 4]
```diff
@@ -78,17 +78,49 @@ template<typename MatType>
...
+    output = input;
+  }
+  else
+  {
+    mask.randu(input.n_rows, input.n_cols);
+    #pragma omp parallel for collapse(2)
+    for (size_t i = 0; i < input.n_rows; ++i) {
+      for (size_t j = 0; j < input.n_cols; ++j) {
+        mask(i, j) = (mask(i, j) > this->ratio) ? 1.0 : 0.0;
+      }
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/dropout_impl.hpp:100]:
```suggestion
    for (size_t i = 0; i < input.n_rows; ++i)
    {
      for (size_t j = 0; j < input.n_cols; ++j)
      {
        mask(i, j) = (mask(i, j) > this->ratio) ? 1.0 : 0.0;
      }
```

File: src/mlpack/methods/ann/layer/dropout_impl.hpp [hunk 5]
```diff
@@ -78,17 +78,49 @@ template<typename MatType>
...
-    // 'ratio'.
     mask.randu(input.n_rows, input.n_cols);
-    mask.transform([&](double val) { return (val > ratio); });
-    output = input % mask * scale;
+    #pragma omp parallel for collapse(2)
+    for (size_t i = 0; i < input.n_rows; ++i) {
+      for (size_t j = 0; j < input.n_cols; ++j) {
+        mask(i, j) = (mask(i, j) > this->ratio) ? 1.0 : 0.0;
+      }
+    }
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/dropout_impl.hpp:122]: Could you update it as above to me in mlpack style ?

**Review by shrit — CHANGES_REQUESTED** [review]

File: src/mlpack/methods/ann/layer/dropout_impl.hpp [hunk 6]
```diff
@@ -78,17 +78,53 @@ template<typename MatType>
...
+    {
+      for (size_t j = 0; j < input.n_cols; ++j)
+      {
+        mask(i, j) = (mask(i, j) > this->ratio) ? 1.0 : 0.0;
+      }
+    }
+    output = input % mask * this->scale;
+  }
+}
+
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/dropout_impl.hpp:107]: @MarkFischinger It looks to me as if it is the same implementation when using Armadillo or Bandicoot.
It this is the case, then it is fine, no worries, I would just keep one of them, and remove SFINAE and the other overload since we no longer use the transform function.
Let me know if I overlooked something

**Review by shrit — CHANGES_REQUESTED** [review]

**Comment by MarkFischinger** [comment]
The updated Forward Dropout Layer implementation shows approximately  $23.2$ % increase in speed on the benchmark :)

Here are the benchmarks for the OpenMP Find and Fill on an MNIST Benchmark:
```
756/756 [====================================================================================================] 100% - 69.8892s/epoch; 92ms/step; loss: 9.42397
Validation loss: 10191.9.
Time elapsed: 72218ms
Accuracy: train = 70.4048%,      valid = 70.1429%
```

Original Version
```
756/756 [====================================================================================================] 100% - 89.117s/epoch; 117ms/step; loss: 9.42397
Validation loss: 10191.9.
Time elapsed: 91523ms
Accuracy: train = 70.4048%,      valid = 70.1429%
```

**Comment by shrit** [comment]
@MarkFischinger this is still failing can you be sure that this is compiling on your machine ?

**Comment by shrit** [comment]
@MarkFischinger can you checkout to the master and pull out the latest version, then you go back to this branch and merge it with the master. It will show you the conflict it has with the master so you can resolve it and push it here again.

**Review by shrit — APPROVED** [review]
@rcurtin nothing to add on my end, let us get this one merged

File: src/mlpack/methods/ann/layer/dropout_impl.hpp [hunk 7]
```diff
@@ -77,18 +77,22 @@ DropoutType<MatType>::operator=(DropoutType&& other)
...
 {
-  // The dropout mask will not be multiplied in testing mode.
   if (!this->training)
   {
     output = input;
   }
   else
   {
-    // Scale with input / (1 - ratio) and set values to zero with probability
-    // 'ratio'.
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/dropout_impl.hpp:88]: @MarkFischinger do you think you can re-add these comments?  I think they are useful to keep.
